### PR TITLE
Take monster reference in M_ClearSquares

### PR DESCRIPTION
**Pull request by ephphatha**

### CODE REVIEW

**Review by AJenbo — DISMISSED** [review]
Do you think something similar could be done with `RemovePlrFromMap()`? Doing so would solve a chain of player references.

It also currently searches the whole level, which could probably also be limited to an area around the last known position (in case of knockback).

File: Source/msg.cpp [hunk 1]
```diff
@@ -2404,10 +2404,11 @@ void DeltaLoadLevel()
 			if (deltaLevel.monster[i]._mx == 0xFF)
 				continue;
 
-			M_ClearSquares(i);
+			// Should this be Monsters[ActiveMonsters[i]]? this loop uses ActiveMonsterCount, not MAXMONSTERS
```
[the remark this answers is not in the document]
Comment (AJenbo) [reply Source/msg.cpp:2407]: So the proper var to use is `totalmonsters`

Also it looks like we should be calculating ActiveMonsterCount and ActiveMonsters here.

**Comment by ephphatha** [comment]
> Do you think something similar could be done with `RemovePlrFromMap()`? Doing so would solve a chain of player references.
> 
> It also currently searches the whole level, which could probably also be limited to an area around the last known position (in case of knockback).

FixPlrWalkTags is the equivalent function for players, looks like RemovePlrFromMap was an overly defensive version that can now be deduplicated. They can be combined and use the exact same approach, I'll open a PR if I get it done tonight.

File: Source/msg.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (AJenbo) [reply Source/msg.cpp:2404]: could you rework this in to a comment for line 2399:

`BUGFIX: Iterate over all monsters and re calculate ActiveMonsterCount`

[the remark this answers is not in the document]
Comment (ephphatha) [reply Source/msg.cpp]: This function is called pretty much immediately after loadlevel spawns all monsters so it should only be applying changes in the form of marking monsters as dead/to be deleted. Not 100% sure how this interacts with golems/the last 4 active monsters? Don't think it's harmful to just iterate over the whole array though so I've just done that in the last commit.

File: Source/msg.cpp [hunk 2]
```diff
@@ -2397,12 +2397,12 @@ void DeltaLoadLevel()
 	uint8_t localLevel = GetLevelForMultiplayer(*MyPlayer);
 	DLevel &deltaLevel = GetDeltaLevel(localLevel);
 	if (leveltype != DTYPE_TOWN) {
-		for (int i = 0; i < ActiveMonsterCount; i++) {
+		for (int i = 0; i < MaxMonsters; i++) {
 			if (deltaLevel.monster[i].position.x == 0xFF)
 				continue;
```
Comment (AJenbo) [inline Source/msg.cpp:2403]: Isn't this also needed?
```suggestion
		ActiveMonsterCount = 0;
		for (int i = 0; i < MaxMonsters; i++) {
			if (deltaLevel.monster[i].position.x == 0xFF)
				continue;

			ActiveMonsterCount++;
```

Comment (AJenbo) [reply Source/msg.cpp:2403]: in conclusion adding `ActiveMonsters[]` would have made the function more correct/safe and only iterate over the spawned monsters.

Using `MaxMonsters` also make the function correct, and all unspawned monsters will simply be skipped as they are `0xFF` in the delta array.

**Review by AJenbo — APPROVED** [review]

**Comment by AJenbo** [comment]
GitHub buttons are broken again :(

File: Source/msg.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (AJenbo) [reply Source/msg.cpp]: Golems should be the first 4

**Comment by AJenbo** [comment]
Auto-merge should happen in about 3 hours if i'm reading AppVeyor right

**Comment by AJenbo** [comment]
merged manually via https://github.com/diasurgical/devilutionX/commit/fce4d23ba8188262852f5bb21e98b4222f7bff02 and https://github.com/diasurgical/devilutionX/commit/2da2ea7b0388c450a87f022e3fdf7b03b1579fd4. No thanks to GitHub :(

**Pull request closed without merging**